### DISCUSSION TITLE
Using cy.task to test GraphQL API errors

### DISCUSSION THREAD

**Discussion by Fgerthoffert** [cypress-io/cypress#16521] [Questions and Help] [answered]
Hi,

I'm using `cy.task()` to test a GraphQL API directly from Node. 

It's working perfectly well for "regular" calls, but I'm also trying to validate exceptions being returned when an incorrect payload is submitted.

What would be the best way to deal with these with cy.task() ?

<img width="450" alt="Screen Shot 2021-05-14 at 10 48 18 AM" src="https://user-images.githubusercontent.com/5667028/118287930-ea62a600-b4a1-11eb-8d57-af7e8cbb400f.png">

The message `'' is not a valid node path` is what I'm trying to actually validate here.

Thanks,

**Comment by bahmutov** [comment, marked as answer]
Can you provide more details, please? Like source code and test code for a minimal example that we can run. Right now no one can say for sure what is going on from the above screenshot.

**Reply by Fgerthoffert** [reply]
True, sorry for that.

Here's a code sample:

**The test:**
```typescript
it('gets a non existing user', () => {
    cy.task('apolloNode', {
        baseUrl: Cypress.config().baseUrl,
        authMethod: { username: 'root', password: Cypress.env('SUPER_USER_PASSWORD') },
        query: gql`
            {
                admin {
                    userAdmin {
                        user(userName: "noob") {
                            name
                        }
                    }
                }
            }
        `,
    }).then(async (response: any) => {
        cy.log(JSON.stringify(response))
        expect(response.data.admin.userAdmin).to.exist
        expect(response.data.admin.userAdmin.user).to.be.null
    })
})
```

The codebase is available here: https://github.com/Jahia/graphql-core/blob/master/tests/cypress/integration/api/admin/user.spec.ts

In the above test, if the user doesn't exist we return null, but assume the API would return an actual error we would like to verify.

**The response to be tested**:
```json
{
  "data": {
    "admin": {
      "userAdmin": {
        "user": null,
      }
    }
  },
  "errors": [
    {
      "message": "An Error we would like to validate",
      "path": null,
      "extensions": null
    }
  ]
}
```

Thanks

**Reply by Fgerthoffert** [reply]
So my apologies for sending this question prematurely, I'm going to respond to myself here.

Using `errorPolicy: 'all'` does not throw an exception and `Both data and error.graphQLErrors are populated, enabling you to render both partial results and error information.` 
Source: https://www.apollographql.com/docs/react/data/error-handling/

```
return new ApolloClient({
    cache: new InMemoryCache(),
    link: from([(authLink as unknown) as ApolloLink, errorLink, httpLink]),
    defaultOptions: {
        query: {
            errorPolicy: 'all',
            fetchPolicy: 'network-only',
        },
    },
})
```

Therefore the response available to `cy.tasks()` contains the data I wanted to validate in `then()`.
